packages/react-dom/src/__tests__/ReactDOMFiber-test.js

```diff
@@ -989,6 +989,40 @@ describe('ReactDOMFiber', () => {
     }
   });
 
+  it('does not fire mouseEnter twice', () => {
+    const ops = [];
+    let target = null;
+
+    function simulateMouseMove(from, to) {
+      if (from) {
+        from.dispatchEvent(
+          new MouseEvent('mouseout', {
+            bubbles: true,
+            cancelable: true,
+            relatedTarget: to,
+          }),
+        );
+      }
+      if (to) {
+        to.dispatchEvent(
+          new MouseEvent('mouseover', {
+            bubbles: true,
+            cancelable: true,
+            relatedTarget: from,
+          }),
+        );
+      }
+    }
+
+    ReactDOM.render(
+      <div onMouseEnter={() => ops.push('enter')} ref={n => (target = n)} />,
+      container,
+    );
+
+    simulateMouseMove(target.parentElement, target);
```

Comment:
> mouseout is also delivered to an element if the cursor enters a child element, because the child element obscures the visible area of the element.

-- https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseout_event

> it MUST be dispatched when the pointer device moves from an element onto the boundaries of one of its descendent elements.

-- https://w3c.github.io/uievents/#event-type-mouseout